packages/eslint-plugin/src/rules/no-unnecessary-template-expression.ts

```diff
@@ -51,21 +51,29 @@ export default createRule<[], MessageId>({
     function isUnderlyingTypeString(
       expression: TSESTree.Expression,
     ): expression is TSESTree.Identifier | TSESTree.StringLiteral {
-      const type = getConstrainedTypeAtLocation(services, expression);
+      const checker = services.program.getTypeChecker();
+      const { constraintType } = getConstraintInfo(
+        checker,
+        services.getTypeAtLocation(expression),
+      );
+
+      if (constraintType == null) {
```

Comment:
Could you ensure there's a test for this case, please? Incidentally, ensuring that we handling unconstrained generics properly is sort of the whole reasoning behind the change to `getConstraintInfo`, so it's working as intended hehe 😁 

I probably should have added this context to the issue description - I'll go do so now!